spring-kafka/src/main/java/org/springframework/kafka/requestreply/ValidatingReplyingKafkaTemplate.java

```diff
@@ -0,0 +1,77 @@
+package org.springframework.kafka.requestreply;
+
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.common.header.Header;
+import org.springframework.kafka.core.ProducerFactory;
+import org.springframework.kafka.listener.GenericMessageListenerContainer;
+import org.springframework.kafka.requestreply.CorrelationKey;
+import org.springframework.kafka.requestreply.ReplyingKafkaTemplate;
+import org.springframework.kafka.support.KafkaHeaders;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * This is a {@link ReplyingKafkaTemplate} that adds a simple validation semantic so it can
+ * take multiple responses and choose the first one that matches the validation condition.
+ * <p>
+ * The use case for this would be providing a farm of topic responders which are decoupled
+ * from the calling service and the calling service does not know who would respond and when,
+ * but knows some property of the response to consider it <em>valid</em>.
+ * <p>
+ * This can be explained using a dinner party analogy:
+ * <ol>
+ * <li>0:00 Kid: Does anyone know what the answers are to the square root of 144 and 2+2?
+ * <li>0:01 Uncle 1: 13, 5
+ * <li>0:02 Uncle 2: 12, 4
+ * <li>0:05 Kid: okay I gathered a few answers, 
+ * <li>0:05 Kid: Filter out who can't answer 2+2
+ * <li>0:05 Kid: The proper answer is 12, 4
+ * <li>0:06 Uncle 3: 12, 4
+ * <li>0:06 Kid: Sorry uncle 3 you're too slow, so I am ignoring you
+ * </ul>
+ */
+public class ValidatingReplyingKafkaTemplate<K, V, R> extends ReplyingKafkaTemplate<K, V, R> {
+
+    private final Function<ConsumerRecord<K, R>, Boolean> validationMethod;
```

Comment:
You could use a `BiPredicate` here.

Reply:
> 
> 
> Right, but you can pull it up as a `protected final` method to the super class which has access to `futures`.

That would mean I would have to change the existing code which is what I am trying to avoid.  But okay

Reply:
Of course, but there's no other way to do it aside from keeping your own copy of `futures`. Which is fine, if you just want to do that locally, but in the framework, we need to do it properly.